perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -2048,231 +2226,315 @@ private boolean isThisLightweightSynchronization(PerunSession sess, Group group)
+		// Set correct member Status
+		// If the member has expired or disabled status, try to expire/validate him (depending on expiration date)
+		if (richMember.getStatus().equals(Status.DISABLED) || richMember.getStatus().equals(Status.EXPIRED)) {
+			Date now = new Date();
+			Attribute membershipExpiration = getPerunBl().getAttributesManagerBl().getAttribute(sess, richMember, AttributesManager.NS_MEMBER_ATTR_DEF + ":membershipExpiration");
+			if(membershipExpiration.getValue() != null) {
+				try {
+					Date currentMembershipExpirationDate = BeansUtils.getDateFormatterWithoutTime().parse((String) membershipExpiration.getValue());
+					if (currentMembershipExpirationDate.before(now)) {
+						// Disabled members which are after expiration date will be expired
+						if (richMember.getStatus().equals(Status.DISABLED)) {
 							try {
-								perunBl.getMembersManagerBl().validateMember(sess, richMember);
-								log.info("Switching member id {} to VALID state, due to expiration {}.", richMember.getId(), (String) membershipExpiration.getValue());
-								log.debug("Switching member to VALID state, additional info: membership expiration date='{}', system now date='{}'", currentMembershipExpirationDate, now);
-							} catch (WrongAttributeValueException e) {
-								log.error("Error during validating member id {}, exception {}", richMember.getId(), e);
-							} catch (WrongReferenceAttributeValueException e) {
-								log.error("Error during validating member id {}, exception {}", richMember.getId(), e);
+								perunBl.getMembersManagerBl().expireMember(sess, richMember);
+								log.info("Switching member id {} to EXPIRE state, due to expiration {}.", richMember.getId(), (String) membershipExpiration.getValue());
+								log.debug("Switching member to EXPIRE state, additional info: membership expiration date='{}', system now date='{}'", currentMembershipExpirationDate, now);
+							} catch (MemberNotValidYetException e) {
+								log.error("Consistency error while trying to expire member id {}, exception {}", richMember.getId(), e);
 							}
 						}
-					} catch (ParseException ex) {
-						log.error("Group synchronization: memberId {} expiration String cannot be parsed, exception {}.",richMember.getId(), ex);
+					} else {
+						// Disabled and expired members which are before expiration date will be validated
+						try {
+							perunBl.getMembersManagerBl().validateMember(sess, richMember);
+							log.info("Switching member id {} to VALID state, due to expiration {}.", richMember.getId(), (String) membershipExpiration.getValue());
+							log.debug("Switching member to VALID state, additional info: membership expiration date='{}', system now date='{}'", currentMembershipExpirationDate, now);
+						} catch (WrongAttributeValueException e) {
+							log.error("Error during validating member id {}, exception {}", richMember.getId(), e);
```

Comment:
It says it runs synchronously. But wasn't this problem before my changes too? I did not change this algorithm, I just moved it to this method, because updateExistingMembersWhileSynchronization method was too long.